20_beta_diversity.Rmd

```diff
@@ -273,50 +273,66 @@ variable_names <- c("ClinicalStatus", "Gender", "Age")
 # Apply relative transform
 enterotype <- transformSamples(enterotype, method = "relabundance")
 
-# Get assay
-assay <- t(assay(enterotype, "relabundance"))
-# Get colData
-coldata <- colData(enterotype)
-
 # Create a formula
 formula <- as.formula(paste0("assay ~ ", str_c(variable_names, collapse = " + ")) )
 
 # # Perform RDA
-rda <- rda(formula, data = coldata, scale = TRUE, na.action = na.exclude)
-
-# Initialize list for p-values
-rda_info <- list()
-# Name for storing the result
-variable_name <- "all"
-# Calculate and store p-value, and other information
-rda_info[[variable_name]] <- c(constrained = rda$CCA$tot.chi, 
-                               unconstrainded = rda$CA$tot.chi, 
-                               proportion = rda$CCA$tot.chi/rda$CA$tot.chi, 
-                               p_value = anova.cca(rda)["Model", "Pr(>F)"] )
-
-# Loop through variables
-permutations <- 999
-for( variable_name in variable_names ){
-    # Create a formula
-    formula <- as.formula(paste0("assay ~ ", variable_name) )
-    # Perform RDA
-    rda_temp <- rda(formula, data = coldata, scale = TRUE, na.action = na.exclude)
-    # Add Info to list
-    rda_info[[variable_name]] <- c(constrained = rda_temp$CCA$tot.chi, 
-                                   unconstrainded = rda_temp$CA$tot.chi, 
-                                   proportion = rda_temp$CCA$tot.chi/rda$CA$tot.chi, 
-                                   p_value = anova.cca(rda_temp, permutations = permutations
-                                                       )["Model", "Pr(>F)"] )
-}  
-# Convert into data.frame
-rda_info <- t(as.data.frame(rda_info))
-rda_info_clean <- rda_info
-# Adjust names
-colnames(rda_info_clean) <- 
-    c("Explained by variables", "Unexplained by variables", "Proportion expl by vars", 
-      paste0("P-value (PERMANOVA ", permutations, " permutations)") )
-# Print info
-kable(rda_info_clean)
+rda <- calculateRDA(enterotype, assay_name = "relabundance",
+                    formula = formula, distance = "bray", na.action = na.exclude)
+# Get the rda object
+rda <- attr(rda, "rda")
+# Calculate p-value and variance for whole model
+set.seed(436)
+permanova <- anova.cca(rda, permutations = 999)
+# Create a data.frame for results
+rda_info <- as.data.frame(permanova)["Model", ]
+
+# Calculate p-value and variance for each variable
+# by = "margin" --> the order or variables does not matter
+set.seed(4585)
+permanova <- anova.cca(rda, by = "margin",  permutations = 999)
+# Add results to data.frame
+rda_info <- rbind(rda_info, permanova)
+
+# Add info about total variance
+rda_info[ , "Total variance"] <- rda_info["Model", 2] +
+    rda_info["Residual", 2]
+
+# Add info about explained variance
+rda_info[ , "Explained variance"] <- rda_info[ , 2] / 
+    rda_info[ , "Total variance"]
+
```

Comment:
Just wondering if we could standarize these steps even better and incorporate them into the function `calculateRDA`? While maintaining some standarization among the similarly named runRDA, runMDS, runPCA etc. functions (or creating another wrapper with a different name). The point being, if this is an essential and common task that comes up all the time in taxonomic analyses, why not streamline it.

Reply:
In some sense I symphatize the idea that users calculate everything themselves. 

But for naive users that might be too much, and for experienced users that could cause unnecessary work load as they would just like to not loose time and avoid possibilities for manual errors. And documentation could also show the more detailed ways or link to material that shows.

Reply:
Yes, there are advantages and I think this could be standardized